airbyte-cdk/java/airbyte-cdk/db-sources/src/main/java/io/airbyte/cdk/integrations/debezium/internals/DebeziumShutdownProcedure.java

```diff
@@ -48,29 +49,35 @@ public DebeziumShutdownProcedure(final LinkedBlockingQueue<T> sourceQueue,
 
   private Runnable transfer() {
     return () -> {
-      while (!sourceQueue.isEmpty() || !MoreBooleans.isTruthy(publisherStatusSupplier.get())) {
+      while (!hasEngineShutDown()) {
         try {
-          T event = sourceQueue.poll(10, TimeUnit.SECONDS);
-          if (event != null) {
-            targetQueue.put(event);
-          }
+          targetQueue.put(sourceQueue.poll(10, TimeUnit.SECONDS));
         } catch (final InterruptedException e) {
           Thread.currentThread().interrupt();
           throw new RuntimeException(e);
         }
       }
+      // At this point, the engine has shut down, and therefore nothing will be writing to the queue any
+      // more. Empty it.
+      sourceQueue.drainTo(targetQueue);
     };
   }
 
+  private boolean hasEngineShutDown() {
+    return MoreBooleans.isTruthy(publisherStatusSupplier.get());
+  }
+
   private void initiateTransfer() {
     executorService.execute(transfer());
   }
 
-  public LinkedBlockingQueue<T> getRecordsRemainingAfterShutdown() {
+  public ArrayList<T> getRecordsRemainingAfterShutdown() {
```

Comment:
The return value is created by copying the contents of the underlying `LinkedBlockingQueue` which is thread safe.

Reply:
This is actually a bug that I introduced: because `computeNext` can be called multiple times and all but the first of these records will be lost. See https://github.com/airbytehq/airbyte/pull/31924#discussion_r1374912521